What's new in 3.0 (#14250)
aspnet/AspNetCore.Docs: aspnetcore/release-notes/aspnetcore-3.0.md

---
title: What's new in ASP.NET Core 3.0
author: rick-anderson
description: Learn about the new features in ASP.NET Core 3.0.
ms.author: riande
ms.custom: mvc
ms.date: 12/18/2018
uid: aspnetcore-3.0
---

> https://docs.microsoft.com/en-us/windows-server/security/kerberos/kerberos-authentication-overview

Reviewer: @Rick-Anderson Is this link supposed to go somewhere in the doc?

Author: No, it's a reminder at the end I have a bunch of links that need  a global removal of `/https://docs.microsoft.com/en-us`

---
title: What's new in ASP.NET Core 3.0
author: rick-anderson
description: Learn about the new features in ASP.NET Core 3.0.
ms.author: riande
ms.custom: mvc
ms.date: 09/17/2019
uid: aspnetcore-3.0
---
# What's new in ASP.NET Core 3.0

This article highlights the most significant changes in ASP.NET Core 3.0 with links to relevant documentation.

## ASP.NET Core 3.0 only runs on .NET Core 3.0

> As of ASP.NET Core 3.0, .NET Framework is no longer a supported target framework. Projects targeting .NET Framework can continue in a fully supported fashion using the [.NET Core 2.1 LTS release](https://www.microsoft.com/net/download/dotnet-core/2.1). ASP.NET Core 2.1 is supported until August 21, 2021. For more information, see [this GitHub announcement](https://github.com/aspnet/Announcements/issues/324).

Author: I added this but would like approval by @DamianEdwards

> As of ASP.NET Core 3.0, .NET Framework is no longer a supported target framework. Projects targeting .NET Framework can continue in a fully supported fashion using the [.NET Core 2.1 LTS release](https://www.microsoft.com/net/download/dotnet-core/2.1). ASP.NET Core 2.1 is supported until August 21, 2021. For more information, see [this GitHub announcement](https://github.com/aspnet/Announcements/issues/324).

Reviewer 2: Fair enough. If we think it's more complicated than is necessary in the What's New post, that also makes sense to me.

* [Newtonsoft.Json](https://www.nuget.org/packages/Newtonsoft.Json/) (Json.NET). To add Json.NET to ASP.NET Core 3.0, see [Add Newtonsoft.Json-based JSON format support](xref:web-api/advanced/formatting#add-newtonsoftjson-based-json-format-support). ASP.NET Core 3.0 introduces `System.Text.Json` for reading and writing JSON. For more information, see [New JSON serialization](#json) in this document.
* [Entity Framework Core](/ef/core/)

For a complete list of assemblies removed from the shared framework, see [Assemblies being removed from Microsoft.AspNetCore.App 3.0](https://github.com/aspnet/AspNetCore/issues/3755). For more information on the motivation for this change, see [Breaking changes to Microsoft.AspNetCore.App in 3.0](https://github.com/aspnet/Announcements/issues/325) and [A first look at changes coming in ASP.NET Core 3.0](https://devblogs.microsoft.com/aspnet/a-first-look-at-changes-coming-in-asp-net-core-3-0/).

<a name="json"></a>

### New JSON serialization

ASP.NET Core 3.0 includes `System.Text.Json`:

* Reads and writes JSON asynchronously.
* Is optimized for UTF-8 text.
* Typically higher performance than `Newtonsoft.Json`.

## gRPC

> See <xref:grpc/index>.

Reviewer 3: Yup, that's a gRPC link all right.

Do you want any other feedback from me? 😋

Reviewer 3: More seriously I think a one paragraph description of gRPC is and what we're shipping, like Razor components has, would be useful.

Author: Can you draft that? You can leave it as a comment and I'll wordsmith and commit.

Reviewer 3: [gRPC](https://grpc.io/) is a popular, high-performance RPC (remote procedure call) framework that offers an opinionated contract-first approach to API development. It uses modern technologies such as HTTP/2 for transport, and Protocol Buffers as the interface description language and binary serialization format. gRPC provides features such as authentication, bidirectional streaming and flow control, and cancellation and timeouts.

gRPC functionality in ASP.NET Core 3.0 includes:

* [Grpc.AspNetCore](https://www.nuget.org/packages/Grpc.AspNetCore) &ndash; An ASP.NET Core framework for hosting gRPC services. gRPC on ASP.NET Core integrates with standard ASP.NET Core features like logging, dependency injection (DI), authentication and authorization.
* [Grpc.Net.Client](https://www.nuget.org/packages/Grpc.Net.Client) &ndash; A gRPC client for .NET Core that builds upon the familiar `HttpClient`.
* [Grpc.Net.ClientFactory](https://www.nuget.org/packages/Grpc.Net.ClientFactory) &ndash; gRPC client integration with `HttpClientFactory`.

## New Razor features

The following list contains new Razor features:

> * [@attribute](xref:mvc/views/razor#attribute) &ndash; The `@attribute` directive applies the given attribute to the class of the generated page or view. For example, `@Attribute [Authorize]`.

Reviewer 4: `@~A~attribute

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddAuthentication(NegotiateDefaults.AuthenticationScheme)
        .AddNegotiate();
    // Other service configuration removed.
}

public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
    app.UseAuthentication();
    // Other app configuration removed.
}
```

Host requirements:

* Windows hosts must have [Service Principal Names](/windows/win32/ad/service-principal-names) (SPNs) added to the user account hosting the app.
* Linux and macOS machines must be joined to the domain.

  * SPNs must be created for the web process.
  * [Keytab files](https://blogs.technet.microsoft.com/pie/2018/01/03/all-you-need-to-know-about-keytab-files/) must be generated and configured on the host machine.

See <xref:security/authentication/windowsauth> for more information.

> ## Template changes

Reviewer 4: Should this section also include info about the Razor Class Library change? As by default it's now a blazor class library by default and there is an optional CLI flag `-support-pages-and-views`?

Reviewer 5: Done

The web UI templates (Razor Pages, MVC with controller and views) have the following removed:

* The cookie consent UI is no longer included. To enable the cookie consent feature in an ASP.NET Core 3.0 template generated app, see <xref:security/gdpr>.
* Scripts and related static assets are now referenced as local files instead of using CDNs. For more information, see [this GitHub issue](https://github.com/aspnet/AspNetCore.Docs/issues/14350).

The Angular template updated to use Angular 8.

## Razor components

Razor components are self-contained chunks of user interface (UI), such as a page, dialog, or form. Razor components are normal .NET classes that define UI rendering logic and client-side event handlers. You can create rich interactive web apps without JavaScript. 

Razor components are typically authored using Razor syntax, a natural blend of HTML and C#. Razor components are similar to Razor Pages and MVC views in that they both use Razor. Unlike pages and views, which are based on a request-response model, components are used specifically for handling UI composition. 

For more information, see <xref:blazor/index>.

## SignalR

> See [Update SignalR code](xref:migration/22-to-30#update-signalr-code) for migration instructions.

Reviewer 2: I might suggest a small blurb about the move to System.Text.Json (since it's probable some SignalR customers would hit issues). I know there's a reference to it above, but this would call out the  I can't do a suggestion because the suggestion would include code fences, but you could edit this comment, copy out the below and paste it in to "apply" this suggestion.

Another option would be to put this blurb in a new heading section of https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/signalr/configuration.md and then link to it here.

---

SignalR now uses System.Text.Json to implement the JSON Hub Protocol. There are some serialization scenarios previously supported by Newtonsoft.Json which are not currently supported in System.Text.Json. If your server app is affected by one of those scenarios, you can revert to the Newtonsoft.Json-based protocol package by adding a reference to the `Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson` NuGet package and chaining a call to `AddNewtonsoftJson()` to your `AddSignalR` call in `ConfigureServices`:

```csharp
services.AddSignalR()
    .AddNewtonsoftJsonProtocol() // Revert to Newtonsoft.Json protocol implementation
```

Similarly, to revert the .NET client to the Newtonsoft.Json-based protocol implementation, add a reference to the `Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson` NuGet package and chaining a call to `AddNewtonsoftJson()` to your `HubConnectionBuilder` calls:

```csharp
var connection = new HubConnectionBuilder()
    .WithUrl("...")
    .AddNewtonsoftJsonProtocol() // Revert to Newtonsoft.Json protocol implementation
    .Build();
```

Author: @anurse  that's in the migration doc. I did a a pointer to that specifically.

* Synchronous IO APIs, such as `HttpReqeuest.Body.Read`, are a common source of thread starvation leading to app crashes. In 3.0, `AllowSynchronousIO` is disabled by default.

For more information, see <xref:migration/22-to-30>.

## Hosting EventSourceProvider new request counters

The Hosting EventSourceProvider (Microsoft.AspNetCore.Hosting) emits the following request counters:

* `requests-per-second`
* `total-requests`
* `current-requests`
* `failed-requests`

## Endpoint routing

See https://github.com/aspnet/AspNetCore.Docs/issues/14291

## Health Checks

Health Checks use endpoint routing with the Generic Host. In `Startup.Configure`, call `MapHealthChecks` on the endpoint builder with the endpoint URL or relative path:

```csharp
app.UseEndpoints(endpoints =>
{
    endpoints.MapHealthChecks("/health");
});
```

Health Checks endpoints can:

* Specify one or more permitted hosts/ports.
* Require authorization.
* Require CORS.

For more information, see the following articles:

* <xref:migration/22-to-30#health-checks>
* <xref:host-and-deploy/health-checks>

## Pipes on HttpContext

> TODO by PU

Reviewer 6: LGTM.

> TODO by PU

Reviewer 7: This is fine with me too. It might be worth calling out these are analogues to the HttpRequest/Response.Body Streams, but I could go either way on that.

<!-- indirectly related, https://github.com/dotnet/docs/pull/14414 won't be published by 9/23  -->

## Worker service and SDK

See https://github.com/aspnet/AspNetCore.Docs/issues/14269

## Forwarded Headers Middleware improvements

In previous versions, calling <xref:Microsoft.AspNetCore.Builder.HstsBuilderExtensions.UseHsts*> and  <xref:Microsoft.AspNetCore.Builder.HttpsPolicyBuilderExtensions.UseHttpsRedirection*> were problematic when deployed to an Azure Linux or behind any reverse proxy other than IIS. The fix for previous versions is documented in [Forward the scheme for Linux and non-IIS reverse proxies](xref:host-and-deploy/proxy-load-balancer#forward-the-scheme-for-linux-and-non-iis-reverse-proxies).

In ASP.NET Core 3.0, this problem has been fixed. The host enables the [Forwarded Headers Middleware](xref:host-and-deploy/proxy-load-balancer#forwarded-headers-middleware-options) when the `ASPNETCORE_FORWARDEDHEADERS_ENABLED` environment variable has been set to `true`.

> ## Additional information

Reviewer 7: I like this. This is really nitpicky, but "HTTP/2 is now enabled by default in Kestrel for HTTPS endpoints." is a little more accurate since the protocol is configured on a per-endpoint basis.

> ## Additional information

Reviewer 2: I like that modification. Will update my "suggestion"

<!-- 

> For the complete list of changes, see the [ASP.NET Core 2.2 Release Notes](https://github.com/aspnet/Home/releases/tag/2.2.0).

Reviewer 4: ASP.NET Core 2.2 ?

Author: @mkArtakMSFT  that's commented out. I'm hoping to get the same URL for the 3.0 release.

---
title: What's new in ASP.NET Core 3.0
author: rick-anderson
description: Learn about the new features in ASP.NET Core 3.0.
ms.author: riande
ms.custom: mvc
ms.date: 09/17/2019
uid: aspnetcore-3.0
---
# What's new in ASP.NET Core 3.0

This article highlights the most significant changes in ASP.NET Core 3.0 with links to relevant documentation.

> ## ASP.NET Core 3.0 only runs on .NET Core 3.0

Reviewer 8: Is this really what we want to open with? Having the first thing in this doc be a feature cut seems like a bit of a downer.

Author: Moved to the end. I think most folks know this already.

As of ASP.NET Core 3.0, .NET Framework is no longer a supported target framework. Projects targeting .NET Framework can continue in a fully supported fashion using the [.NET Core 2.1 LTS release](https://www.microsoft.com/net/download/dotnet-core/2.1). ASP.NET Core 2.1 is supported until August 21, 2021. For more information, see [this GitHub announcement](https://github.com/aspnet/Announcements/issues/324).

See [Port your code from .NET Framework to .NET Core](/dotnet/core/porting/) for migration information.

> ## Use the ASP.NET Core shared framework

Reviewer 8: @DamianEdwards Do we want to add content here about why we made this change and how it differs from the previous approaches?

> The [ASP.NET Core 3.0 shared framework](https://natemcmaster.com/blog/2018/08/29/netcore-primitives-2/), contained in the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app), no longer requires an explicit `<PackageReference />` element in the project file. The shared framework is automatically referenced when using the `Microsoft.NET.Sdk.Web` SDK in your project file:

Reviewer 8: Let's not link to third party blog posts from official docs that we don't have control over.

```suggestion
The ASP.NET Core 3.0 shared framework, contained in the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app), no longer requires an explicit `<PackageReference />` element in the project file. The shared framework is automatically referenced when using the `Microsoft.NET.Sdk.Web` SDK in your project file:
```

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
```

## Assemblies removed from the ASP.NET Core shared framework

The most notable assemblies removed from the ASP.NET Core 3.0 shared framework are:

* [Newtonsoft.Json](https://www.nuget.org/packages/Newtonsoft.Json/) (Json.NET). To add Json.NET to ASP.NET Core 3.0, see [Add Newtonsoft.Json-based JSON format support](xref:web-api/advanced/formatting#add-newtonsoftjson-based-json-format-support). ASP.NET Core 3.0 introduces `System.Text.Json` for reading and writing JSON. For more information, see [New JSON serialization](#json) in this document.
* [Entity Framework Core](/ef/core/)

For a complete list of assemblies removed from the shared framework, see [Assemblies being removed from Microsoft.AspNetCore.App 3.0](https://github.com/aspnet/AspNetCore/issues/3755). For more information on the motivation for this change, see [Breaking changes to Microsoft.AspNetCore.App in 3.0](https://github.com/aspnet/Announcements/issues/325) and [A first look at changes coming in ASP.NET Core 3.0](https://devblogs.microsoft.com/aspnet/a-first-look-at-changes-coming-in-asp-net-core-3-0/).

<a name="json"></a>

> ### New JSON serialization

Reviewer 8: Add a mention of the Microsoft.AspNetCore.Mvc.NewtonsoftJson package for folks that still want to use Json.NET.

> ASP.NET Core 3.0 includes `System.Text.Json`:

Reviewer 8: ```suggestion
ASP.NET Core 3.0 now uses `System.Text.Json` by default for JSON serialization.
```

* Reads and writes JSON asynchronously.
* Is optimized for UTF-8 text.
* Typically higher performance than `Newtonsoft.Json`.

## gRPC

See <xref:grpc/index>.

## New Razor features

The following list contains new Razor features:

* [@attribute](xref:mvc/views/razor#attribute) &ndash; The `@attribute` directive applies the given attribute to the class of the generated page or view. For example, `@Attribute [Authorize]`.
* [@implements](xref:mvc/views/razor#implements) &ndash; The `@implements` directive implements an interface for the generated class. For example, `@implements IDisposable`.

> * [@page](xref:mvc/views/razor#page) &ndash; The `@page` directive has different effects depending on the type of the file where it appears. In a *.cshtml* file, `@page` indicates that the file is a Razor Page. In a *.razor* file, `@page` specifies that a Razor component should handle requests directly.

Reviewer 8: `@page` isn't new. We just happen to use the same directive in Razor components. I think this bullet can be removed

}

public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
    app.UseAuthentication();
    // Other app configuration removed.
}
```

Host requirements:

* Windows hosts must have [Service Principal Names](/windows/win32/ad/service-principal-names) (SPNs) added to the user account hosting the app.
* Linux and macOS machines must be joined to the domain.

  * SPNs must be created for the web process.
  * [Keytab files](https://blogs.technet.microsoft.com/pie/2018/01/03/all-you-need-to-know-about-keytab-files/) must be generated and configured on the host machine.

See <xref:security/authentication/windowsauth> for more information.

## Template changes

The web UI templates (Razor Pages, MVC with controller and views) have the following removed:

* The cookie consent UI is no longer included. To enable the cookie consent feature in an ASP.NET Core 3.0 template generated app, see <xref:security/gdpr>.
* Scripts and related static assets are now referenced as local files instead of using CDNs. For more information, see [this GitHub issue](https://github.com/aspnet/AspNetCore.Docs/issues/14350).

The Angular template updated to use Angular 8.

> ## Razor components

Reviewer 8: ```suggestion
## Blazor
```
I'd like to expand on this section and turn it into the Blazor section. The emphasis should be on writing client-side UI with .NET & C# instead of JavaScript. The should look something similar to what we have in the first section of the [Blazor overview](https://docs.microsoft.com/aspnet/core/blazor).

We should also have a nice list of the features support by the Blazor component model:

- Reusable UI components
- Client-side routing
- Component layouts
- Support for dependency injection
- Forms & validation
- Build component libraries with Razor Class Libraries
- JavaScript interop

We also need to be clear that support for Blazor Server apps is available with .NET Core 3.0, but support for Blazor WebAssembly apps is *still in preview*. We actually need to make this clear across the Blazor docs.

@guardrex Is this update something you can help out with?

Reviewer 5: @danroth27 I added content as advised. You probably need to edit it.

Note that the title of what I added could be ...

\#\# Blazor and Razor components

... given what's in the section.

---
title: What's new in ASP.NET Core 3.0
author: rick-anderson
description: Learn about the new features in ASP.NET Core 3.0.
ms.author: riande
ms.custom: mvc
ms.date: 09/19/2019
uid: aspnetcore-3.0
---
# What's new in ASP.NET Core 3.0

This article highlights the most significant changes in ASP.NET Core 3.0 with links to relevant documentation.

## Blazor

Blazor is a new framework in ASP.NET Core for building interactive client-side web UI with .NET:

* Create rich interactive UIs using C# instead of JavaScript.
* Share server-side and client-side app logic written in .NET.
* Render the UI as HTML and CSS for wide browser support, including mobile browsers.

> Using .NET for client-side web development offers the following advantages:

Reviewer 8: I think we can remove this list on the benefits of using .NET, as it is primarily targeted at developers that are new to .NET, which really isn't the audience for this article.

* Write code in C# instead of JavaScript.
* Leverage the existing .NET ecosystem of .NET libraries.
* Share app logic across server and client.
* Benefit from .NET's performance, reliability, and security.
* Stay productive with Visual Studio on Windows, Linux, and macOS.
* Build on a common set of languages, frameworks, and tools that are stable, feature-rich, and easy to use.

Blazor framework supported scenarios:

* Reusable UI components (Razor components)
* Client-side routing
* Component layouts
* Support for dependency injection
* Forms and validation
* Build component libraries with Razor class libraries
* JavaScript interop

For more information, see <xref:blazor/index>.

### Blazor Server

Blazor decouples component rendering logic from how UI updates are applied. Blazor Server provides support for hosting Razor components on the server in an ASP.NET Core app. UI updates are handled over a SignalR connection. Blazor Server is supported in the ASP.NET Core 3.0 release.

> ### Blazor WebAssembly

Reviewer 8: Can we remove this section heading and just have the following paragraph be part of the previous section? I worry that this section heading will give the impression scanning the headings that Blazor WebAssembly is part of .NET Core 3.0 , when it isn't.

> ### Blazor WebAssembly

Reviewer 8: ```suggestion
### Blazor WebAssembly (Preview)
```

> Blazor WebAssembly is a single-page app framework for building interactive client-side web apps with .NET. Blazor WebAssembly uses open web standards without plugins or code transpilation and works in all modern web browsers, including mobile browsers. Blazor WebAssembly is in preview.

Reviewer 8: ```suggestion
Blazor apps can also be run directly in the browser using a WebAssembly-based .NET runtime. Support for Blazor WebAssembly apps is *not* part of ASP.NET Core 3.0, but will be part of a future release.
```

### Razor components

> Razor components are self-contained chunks of user interface (UI), such as a page, dialog, or form. Razor components are normal .NET classes that define UI rendering logic and client-side event handlers. You can create rich interactive web apps without JavaScript.

Reviewer 8: ```suggestion
Blazor apps are built from components. Components are self-contained chunks of user interface (UI), such as a page, dialog, or form. Components are normal .NET classes that define UI rendering logic and client-side event handlers. You can create rich interactive web apps without JavaScript.
```

> Razor components are typically authored using Razor syntax, a natural blend of HTML and C#. Razor components are similar to Razor Pages and MVC views in that they both use Razor. Unlike pages and views, which are based on a request-response model, components are used specifically for handling UI composition.

Reviewer 8: ```suggestion
Components in Blazor are typically authored using Razor syntax, a natural blend of HTML and C#. Razor components are similar to Razor Pages and MVC views in that they both use Razor. Unlike pages and views, which are based on a request-response model, components are used specifically for handling UI composition.
```

{
    await foreach (var item in stream)
    {
        // process content
    }
}
```

.NET client apps can pass either an `IAsyncEnumerable<T>` or `ChannelReader<T>` instance as the `stream` argument of the `UploadStream` Hub method above.

```csharp
async IAsyncEnumerable<string> clientStreamData()
{
    for (var i = 0; i < 5; i++)
    {
        var data = await FetchSomeData();
        yield return data;
    }
    //After the for loop has completed and the local function exits the stream completion is sent.
}

await connection.SendAsync("UploadStream", clientStreamData());
```

JavaScript client apps use the SignalR `Subject` (or an [RxJS Subject](https://rxjs.dev/api/index/class/Subject)) for the `stream` argument of the `UploadStream` Hub method above.

```javascript
let subject = new signalR.Subject();
await connection.send("StartStream", "MyAsciiArtStream", subject);
```

The JavaScript code could use the `subject.next` method to handle strings as they are captured and ready to be sent to the server.

```javascript
subject.next("example");
subject.complete();
```

Using code like the two preceding snippets, real-time streaming experiences can be created.

> ## Use the ASP.NET Core shared framework

Reviewer 8: How about we move this section and the following section on assemblies removed from the shared framework to the bottom of the article?

* [Entity Framework Core](/ef/core/)

For a complete list of assemblies removed from the shared framework, see [Assemblies being removed from Microsoft.AspNetCore.App 3.0](https://github.com/aspnet/AspNetCore/issues/3755). For more information on the motivation for this change, see [Breaking changes to Microsoft.AspNetCore.App in 3.0](https://github.com/aspnet/Announcements/issues/325) and [A first look at changes coming in ASP.NET Core 3.0](https://devblogs.microsoft.com/aspnet/a-first-look-at-changes-coming-in-asp-net-core-3-0/).

<a name="json"></a>

### New JSON serialization

ASP.NET Core 3.0 now uses `System.Text.Json` by default for JSON serialization:

* Reads and writes JSON asynchronously.
* Is optimized for UTF-8 text.
* Typically higher performance than `Newtonsoft.Json`.

To add Json.NET to ASP.NET Core 3.0, see [Add Newtonsoft.Json-based JSON format support](xref:web-api/advanced/formatting#add-newtonsoftjson-based-json-format-support).

## gRPC

> [gRPC](https://grpc.io/):

Reviewer 3: IMO this is a little heavy on bullet points but it's not a hill I'll die on.

### Blazor WebAssembly

Blazor WebAssembly is a single-page app framework for building interactive client-side web apps with .NET. Blazor WebAssembly uses open web standards without plugins or code transpilation and works in all modern web browsers, including mobile browsers. Blazor WebAssembly is in preview.

### Razor components

Razor components are self-contained chunks of user interface (UI), such as a page, dialog, or form. Razor components are normal .NET classes that define UI rendering logic and client-side event handlers. You can create rich interactive web apps without JavaScript.

Razor components are typically authored using Razor syntax, a natural blend of HTML and C#. Razor components are similar to Razor Pages and MVC views in that they both use Razor. Unlike pages and views, which are based on a request-response model, components are used specifically for handling UI composition.

## SignalR

> See [Update SignalR code](xref:migration/22-to-30#update-signalr-code) for migration instructions. [Switch to Newtonsoft.Json](xref:migration/22-to-30#switch-to-newtonsoftjson) has instructions for migrating to `System.Text.Json`.

Reviewer 2: ```suggestion
See [Update SignalR code](xref:migration/22-to-30#update-signalr-code) for migration instructions. SignalR now uses `System.Text.Json` to serialize/deserialize JSON messages. See [Switch to Newtonsoft.Json](xref:migration/22-to-30#switch-to-newtonsoftjson) for instructions to restore the `Newtonsoft.Json`-based serializer.
```

I think just one more sentence explaining the actual *new* thing and why you'd want to view the `Switch to Newtonsoft.Json` doc might help here.

In the JavaScript and .NET Clients for SignalR, support was added for automatic reconnection. By default, the client tries to reconnect immediately and retry after 2, 10, and 30 seconds if necessary. If the client successfully reconnects, it receives a new connection ID. Automatic reconnect is opt-in:

```javascript
const connection = new signalR.HubConnectionBuilder()
    .withUrl("/chatHub")
    .withAutomaticReconnect()
    .build();
```

The reconnection intervals can be specified by passing an array of millisecond-based durations:

```javascript
.withAutomaticReconnect([0, 3000, 5000, 10000, 15000, 30000])
//.withAutomaticReconnect([0, 2000, 10000, 30000]) The default intervals.
```

A custom implementation can be passed in for full control of the reconnection intervals.

> If the reconnection fails after the maximum reconnect interval:

Reviewer 7: ```suggestion
If the reconnection fails after the last reconnect interval:
```

* Registering `DomainRestricted` with the authorization middleware.

```csharp
services
    .AddAuthorization(options =>
    {
        options.AddPolicy("DomainRestricted", policy =>
        {
            policy.Requirements.Add(new DomainRestrictedRequirement());
        });
    });
```

SignalR hubs use [Endpoint Routing](xref:fundamentals/routing). SignalR hub connection was previously done explicitly:

```csharp
app.UseSignalR(routes =>
{
    routes.MapHub<ChatHub>("hubs/chat");
});
```

In the previous version, developers needed to wire up controllers, Razor pages, and hubs in a variety of different places. Explicit connection results in a series of nearly-identical routing segments:

```csharp
app.UseSignalR(routes =>
{
    routes.MapHub<ChatHub>("hubs/chat");
});

app.UseRouting(routes =>
{
    routes.MapRazorPages();
});
```

SignalR 3.0 hubs can be routed via endpoint routing. With endpoint routing, typically all routing can be configured in `UseRouting`:

```csharp
app.UseRouting(routes =>
{
    routes.MapRazorPages();
    routes.MapHub<ChatHub>("hubs/chat");
});
```

ASP.NET Core 3.0 SignalR added:

> * Client-to-server streaming. With client-to-server streaming, server-side methods can take instances of either an `IAsyncEnumerable<T>` or `ChannelReader<T>`. In the following C# sample, the `UploadStream` method on the Hub will receive a stream of strings from the client:

Reviewer 2: Feels like this doesn't need to be a list now that there's only one element. I'm OK either way though.

Reviewer 7: ```suggestion
Client-to-server streaming. With client-to-server streaming, server-side methods can take instances of either an `IAsyncEnumerable<T>` or `ChannelReader<T>`. In the following C# sample, the `UploadStream` method on the Hub will receive a stream of strings from the client:
```

```

The JavaScript code could use the `subject.next` method to handle strings as they are captured and ready to be sent to the server.

```javascript
subject.next("example");
subject.complete();
```

Using code like the two preceding snippets, real-time streaming experiences can be created.

## Generic Host

The ASP.NET Core 3.0 templates use <xref:fundamentals/host/generic-host>. Previous versions used <xref:Microsoft.AspNetCore.Hosting.WebHostBuilder>. Using the .NET Core Generic Host (<xref:Microsoft.Extensions.Hosting.HostBuilder>) provides better integration of ASP.NET Core apps with other server scenarios that are not web specific.

The following two images show the changes made to the template generated *Program.cs* file:

 ![2.2 version](aspnetcore-3.0/_static/2.2host.png)
 ![3.0 version](aspnetcore-3.0/_static/3.0host.png)

In the preceding images:

* The ASP.NET Core 2.2 version is shown first. The red boxes indicate code that has been removed in the 3.0 version.
* The ASP.NET Core 3.0 version is shown second. The green boxes indicate code that has been added. The 3.0 version requires `using Microsoft.Extensions.Hosting;`.

> ## Host configuration and options

Reviewer 9: Cut? This is not novel functionality and it's covered in the docs.

For more information, see the following articles:

* <xref:fundamentals/configuration/index>
* <xref:fundamentals/host/generic-host>

## Kestrel

* Kestrel configuration has been updated for the migration to the Generic Host. In 3.0, Kestrel is configured on the web host builder provided by `ConfigureWebHostDefaults`.
* Connection Adapters have been removed from Kestrel and replaced with Connection Middleware, which is similar to HTTP Middleware in the ASP.NET Core pipeline but for lower-level connections.
* The Kestrel transport layer has been exposed as a public interface in `Connections.Abstractions`.
* Ambiguity between headers and trailers has been resolved by moving trailing headers to a new collection.
* Synchronous IO APIs, such as `HttpReqeuest.Body.Read`, are a common source of thread starvation leading to app crashes. In 3.0, `AllowSynchronousIO` is disabled by default.

For more information, see <xref:migration/22-to-30>.

> ## Hosting EventSourceProvider new request counters

Reviewer 2: ```suggestion
## Request counters
```

Let's remove reference to the event source altogether in the heading. At a minimum, it would need to be `EventSource`, not `EventSourceProvider` (as below 😉). I don't see a need to call out event source at all though. These are counters related to requests, that's sufficient description for a header IMO.

## Health Checks

Health Checks use endpoint routing with the Generic Host. In `Startup.Configure`, call `MapHealthChecks` on the endpoint builder with the endpoint URL or relative path:

```csharp
app.UseEndpoints(endpoints =>
{
    endpoints.MapHealthChecks("/health");
});
```

Health Checks endpoints can:

* Specify one or more permitted hosts/ports.
* Require authorization.
* Require CORS.

For more information, see the following articles:

* <xref:migration/22-to-30#health-checks> <!-- ignore warning per @guardrex -->
* <xref:host-and-deploy/health-checks>

## Pipes on HttpContext

It is now possible to read the request body and write the response body using the <xref:System.IO.Pipelines> API. The <!-- <xref:Microsoft.AspNetCore.Http.HttpRequest.BodyReader> --> `HttpRequest.BodyReader` property provides a <xref:System.IO.Pipelines.PipeReader> that can be used to read the request body. The <!-- <xref:Microsoft.AspNetCore.Http.> --> `HttpResponse.BodyWriter` property provides a <xref:System.IO.Pipelines.PipeWriter> that can be used to write the response body.

<!-- indirectly related, https://github.com/dotnet/docs/pull/14414 won't be published by 9/23  -->

> ## Worker service and SDK

Reviewer 9: Pull in the content from the blog post?

See https://github.com/aspnet/AspNetCore.Docs/issues/14269

## Forwarded Headers Middleware improvements

In previous versions, calling <xref:Microsoft.AspNetCore.Builder.HstsBuilderExtensions.UseHsts*> and  <xref:Microsoft.AspNetCore.Builder.HttpsPolicyBuilderExtensions.UseHttpsRedirection*> were problematic when deployed to an Azure Linux or behind any reverse proxy other than IIS. The fix for previous versions is documented in [Forward the scheme for Linux and non-IIS reverse proxies](xref:host-and-deploy/proxy-load-balancer#forward-the-scheme-for-linux-and-non-iis-reverse-proxies).

> In ASP.NET Core 3.0, this problem has been fixed. The host enables the [Forwarded Headers Middleware](xref:host-and-deploy/proxy-load-balancer#forwarded-headers-middleware-options) when the `ASPNETCORE_FORWARDEDHEADERS_ENABLED` environment variable has been set to `true`.

Reviewer 9: which is set automatically in our container images.

> ## HTTP/2 enabled by default

Reviewer 2: I'm probably a bit biased, but I would love this to be a little closer to the top ;). I think it's a fairly important feature and it brings us up to parity with most of our competitive frameworks/platforms.

Reviewer 6: And put it next to/near the Kestrel section, as it only applies to Kestrel.
